WordPress/src/main/java/org/wordpress/android/util/analytics/service/InstallationReferrerServiceLogic.java

```diff
@@ -0,0 +1,127 @@
+import com.android.installreferrer.api.InstallReferrerStateListener;
+import com.android.installreferrer.api.ReferrerDetails;
+
+import org.wordpress.android.analytics.AnalyticsTracker;
+import org.wordpress.android.ui.prefs.AppPrefs;
+import org.wordpress.android.util.AppLog;
+import org.wordpress.android.util.AppLog.T;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.wordpress.android.util.analytics.service.InstallationReferrerServiceStarter.ARG_REFERRER;
+
+
+/**
+ * Background service to retrieve installation referrer information.
+ */
+
+public class InstallationReferrerServiceLogic {
+    private ServiceCompletionListener mCompletionListener;
+    private Object mListenerCompanion;
+    private InstallReferrerClient mReferrerClient;
+    private Context mContext;
+
+    public InstallationReferrerServiceLogic(Context context, ServiceCompletionListener completionListener) {
+        mCompletionListener = completionListener;
+        mContext = context;
+    }
+
+    public void onDestroy() {
+        AppLog.i(T.UTILS, "installation referrer service destroyed");
+    }
+
+    public void performTask(Bundle extras, Object companion) {
+        mListenerCompanion = companion;
+
+        // if a referrer string has been passed already (coming from com.android.vending.INSTALL_REFERRER receiver),
+        // just send it to tracks
+        if (extras != null && extras.containsKey(ARG_REFERRER)) {
+            Map<String, Object> properties = new HashMap<>();
+            properties.put("install_referrer", extras.getString(ARG_REFERRER));
+            AnalyticsTracker.track(AnalyticsTracker.Stat.INSTALLATION_REFERRER_OBTAINED, properties);
+            // mark referrer as obtained now
+            AppPrefs.setInstallationReferrerObtained(true);
+            stopService();
+            return;
+        }
+
+
+        // if not, try to obtain it from Play Store app connection through the Install Referrer API Library if possible
+        mReferrerClient = InstallReferrerClient.newBuilder(mContext).build();
+        mReferrerClient.startConnection(new InstallReferrerStateListener() {
+            @Override
+            public void onInstallReferrerSetupFinished(int responseCode) {
+                switch (responseCode) {
+                    case InstallReferrerResponse.OK:
+                        // Connection established
+                        try {
+                            AppLog.i(T.UTILS, "installation referrer connected");
+                            ReferrerDetails response = mReferrerClient.getInstallReferrer();
+                            // mark referrer as obtained so we don't try fetching it again
+                            // Citing here:
+                            //  Caution: The install referrer information will be available for 90 days and won't
+                            //  change unless the application is reinstalled. To avoid unecessary API calls in your
```

Comment:
Can we change `unecessary` to `unnecessary`?

Reply:
Wow! Nice catch @theck13 :D it's copied from Google's documentation 😅 
https://developer.android.com/google/play/installreferrer/library
<img width="823" alt="screen shot 2018-09-12 at 17 35 49" src="https://user-images.githubusercontent.com/6597771/45451741-5fa6c900-b6b2-11e8-8aaa-7eb2451d718f.png">

fixed in 91c4c58d4a